tiledb/sm/query/reader.cc

```diff
@@ -3200,30 +3213,175 @@ Status Reader::copy_coordinates(
+  // the range it resides in to create two ranges that skip the omitted cell.
+  // We will store the new cell slabs ranges on `rcs_partitions`. We process
+  // the cells concurrently. To avoid contention on a single vector of
+  // `ResultCellSlab` instances, we will assign a vector to each thread in the
+  // thread pool.
+  const size_t num_rcs_partitions =
+      global_state::GlobalState::GetGlobalState().tbb_threads() > 0 ?
+          global_state::GlobalState::GetGlobalState().tbb_threads() :
+          storage_manager_->compute_tp()->concurrency_level();
+  assert(num_rcs_partitions > 0);
+
+  std::vector<std::vector<ResultCellSlab>> rcs_partitions;
+  rcs_partitions.resize(num_rcs_partitions);
+  const size_t rcs_partition_size =
+      result_cell_slabs->size() / num_rcs_partitions;
+  for (auto& rcs_partition : rcs_partitions) {
+    rcs_partition.reserve(rcs_partition_size);
+  }
+
+  // Apply the query condition to each result cell slab partition.
+  const auto statuses = parallel_for(
+      storage_manager_->compute_tp(),
+      0,
+      num_rcs_partitions,
+      [this, &rcs_partitions, rcs_partition_size, result_cell_slabs, stride](
+          const uint64_t p) {
+        std::vector<ResultCellSlab>* const rcs_partition = &rcs_partitions[p];
+
+        const size_t partition_start = p * rcs_partition_size;
+        const size_t partition_end = (p == (rcs_partitions.size() - 1)) ?
+                                         result_cell_slabs->size() :
+                                         partition_start + rcs_partition_size;
+
+        return apply_query_condition_partition(
+            rcs_partition,
+            partition_start,
+            partition_end,
+            *result_cell_slabs,
+            stride);
+      });
+
+  for (const auto& st : statuses)
+    RETURN_NOT_OK(st);
+
+  // Calculate the total number of result cell slabs.
+  uint64_t num_rcs = 0;
+  for (const auto& rcs_partition : rcs_partitions) {
+    num_rcs += rcs_partition.size();
+  }
+
+  // Concatenate all of the result cell slabs in the partitions
+  // into `result_cell_slabs`.
+  *result_cell_slabs = std::move(rcs_partitions[0]);
+  result_cell_slabs->reserve(num_rcs);
+  for (size_t p = 1; p < rcs_partitions.size(); ++p) {
+    std::vector<ResultCellSlab>* const rcs_partition = &rcs_partitions[p];
+    result_cell_slabs->insert(
+        result_cell_slabs->end(),
+        std::make_move_iterator(rcs_partition->begin()),
+        std::make_move_iterator(rcs_partition->end()));
+  }
+
+  return Status::Ok();
+}
+
+Status Reader::apply_query_condition_partition(
+    std::vector<ResultCellSlab>* const rcs_partition,
+    const size_t partition_start,
+    const size_t partition_end,
+    const std::vector<ResultCellSlab>& result_cell_slabs,
+    const uint64_t stride) {
+  // TODO handle tile == nullptr (empty range)
+
+  for (size_t r = partition_start; r < partition_end; ++r) {
```

Comment:
This needs parallelization. I don't mean this specific loop. I mean the strategy on passing over the result cell slab items to apply the query condition needs parallelization. Also this strategy is very inefficient because:

1) It is not vectorized (i.e., no tight loops over contiguous values in each attribute tile)
2) It makes an enormous number of function calls to `condition_.cmp_cell`, whichin turn makes a lot of other function calls with several switch statements.

I am up for discussing a vectorized approach online.